### ISSUE TITLE
Extendable fetchers

### ISSUE THREAD

**Issue by shajra** [nmattia/niv#300]
Hi.

It occurred to me that there's a whole slew of "enhancement" fetching issues already in the tracker:
- #77 
- #78 
- #79

But this seems like it might lead to a lot of development overhead for Niv.  And honestly, the number of things that people might want to update automatically could be expansive and in some cases niche (consider corporate/private needs).  You can cover the typical stuff, but there will always be incidentals.

So I'm thinking Niv should have an extensible system where it does some call-outs to shell, not terribly unlike what Direnv does.  Just need to settle on a good interface for the inputs/outputs of the functions the user needs to provide.  I know that's like 80% of the work, but I'm not there yet.  I just wanted to lodge this issue as an enhancement request and see what you thought.

Like Direnv, you could continue to have built-in support for the things that the most users will benefit from.  But also, by having community support with these shell functions, you will also get some inspiration for what's most useful, and how to implement it.  One can feed the other.

What do you think?

**Comment by michaelpj** [comment]
I still think the approach I suggested in https://github.com/nmattia/niv/pull/137 was nice for this kind of customization. You feed `sources.nix` an overlay when you import it and you can use that to define extra fetchers.

**Comment by shajra** [comment]
@michaelpj I scanned the PR there.  It's good to know it's there.  I missed that feature.

From what I saw, this seems to address fetching, but not updating, right?  I think in my original issue, I didn't make clear that updating is something that I'm really interested in solving for as well, because it's a large portion of what I find Niv useful for.  Otherwise, I'm just crawling a lot of websites.

But if you're approach covers updating as well, then I'm pretty happy.  Just need to play around with it and see how it works.

**Comment by michaelpj** [comment]
Yes, you're quite right, it only handled fetching! At the moment I think all the updating is done by the Haskell code, which makes it a bit harder to extend in an ad-hoc way.

**Comment by michaelpj** [comment]
Also, to be clear, that was just a suggestion, it's not actually implemented in niv.

**Comment by shajra** [comment]
> Yes, you're quite right, it only handled fetching! At the moment I think all the updating is done by the Haskell code, which makes it a bit harder to extend in an ad-hoc way.

Okay, I think we're converging on the same sentiments.  One reason to use a shell breakout is because it allows us to do just a little bit more... specifically the stuff that's in Haskell for updating.

But I don't want to completely cut Haskell out of the equation.  I think for the common cases, it's perfectly reasonable to not have to call shell at all.  Shell would just be an alternate path.

I'll update the description of this issue to make clear my interest in updating in addition to fetching.

**Comment by shajra** [comment]
@michaelpj since you're involved in the Haskell.nix world, you may appreciate that the interest in this ticket butts up against my interest in better automating something like the management of materializations for a Haskell.nix project.  The materialization scripts are similar in spirit to the scripts I proposed in this issue's description.  Except... Haskell.nix scripts actually modify the filesystem.  I was thinking originally of scripts that didn't modify the filesystem, but output JSON to then be injected back into `sources.json` (possibly with some transformation).  Also, Haskell.nix is kind of two-stage (I think)... one to get the hash of the new materialization, and another to save it to the filesystem.

Ultimately, I want happy-path, push-button updating of Nix projects, with good error reporting when things go wrong.  At that point, have I gone way past the original purview of Niv?  Am I really going done that path of making a different tool entirely that may subsume some of the functionality of Niv?

**Comment by shajra** [comment]
I was also reminded that @expipiplus1 also maintains an application that's in this space... https://github.com/expipiplus1/update-nix-fetchgit

I feel that the approach there of syntactically matching Nix expressions might be too fragile for what I have in mind, though.  I can appreciate the appeal of hitting the 80% case.  I just find myself in the 20% case a lot more with my personal projects (Haskell.nix being one of those situations for me).

**Comment by colemickens** [comment]
I also would like this, and is part of the reason that I cannot adopt `niv` for `nixpkgs-wayland`. `nixpkgs-wayland` contains a script, update.sh` that uses `nix-prefetch` to update packages sources and hashes (**including cargoSha256 and vendorSha256**) in a fully automated, CI-driven fashion.

If `niv` supported extensible fetching/updating, then I could probably wire up `nix-prefetch` and then use `niv` to drive most of the updating process.
